README.md

```diff
@@ -877,31 +877,86 @@ interface FIDAttribution {
 ```ts
 interface INPAttribution {
   /**
-   * A selector identifying the element that the user interacted with for
-   * the event corresponding to INP. This element will be the `target` of the
-   * `event` dispatched.
+   * A selector identifying the element that the user first interacted with
+   * as part of the frame where the INP candidate interaction occurred.
+   * If `interactionTarget` is an empty string, that generally means the
+   * element was removed from the DOM as part of the interaction.
    */
-  eventTarget?: string;
+  interactionTarget: string;
   /**
-   * The time when the user interacted for the event corresponding to INP.
-   * This time will match the `timeStamp` value of the `event` dispatched.
+   * The time when the user first interacted during the frame where the INP
+   * candidate interaction occurred (if more than one interaction occurred
+   * within the frame, only the first time is reported).
    */
-  eventTime?: number;
+  interactionTime: number;
   /**
-   * The `type` of the `event` dispatched corresponding to INP.
+   * The type of interaction. This will be either 'pointer' or 'keyboard'
+   * since those are the only types of interactions considered for INP.
    */
-  eventType?: string;
+  interactionType: 'pointer' | 'keyboard';
+
   /**
-   * The `PerformanceEventTiming` entry corresponding to INP.
+   * If the browser supports the Long Animation Frame API and a
+   * `long-animation-frame` entry is detected that corresponds to the INP
+   * frame, it will be reported here.
    */
-  eventEntry?: PerformanceEventTiming;
+  longAnimationFrameEntry?: PerformanceLongAnimationFrameTiming;
```

Comment:
Multiple LoAFs can overlap with the INP window, so does limiting this to just one make sense?

Reply:
@mmocny I know you said you wouldn't expect an arbitrary length list of LoAFs, but what would you think about an array of LoAFs that intersect with [startTime, processingEnd]?

That's a bit more awkward than a single LoAF, but it has the benefit that it gives developers all the information about what scripts are blocking INP (for both input delay and processing time).

If we only want to expose a single LoAF then I think the reasonable options are to pick either:
1. The longest LoAF (@tunetheweb's original proposal), or
2. The LoAF that intersects with [processingStart, processingEnd]

WDYT?

Reply:
We discussed this a bit internally and agreed that it makes the most sense to now expose an array of LoAF entries, and with my most recent changes that array will include any LoAF entry that intersects with [startTime, processingEnd] for the INP candidate interaction. @mmocny I'll tag you into that part of the code so you can take a lok.

Reply:
The loaf that intersects with processingStart / processingEnd I would consider "the" LoAF for this Interaction, and it feels uniquely important vs just a list of all of them.

Maybe if we returned the list in reverse chronological order (which matches order of importance) that would be fine.

---

The one single preceding LoAF might well be useful for debugging input delay.

However, an long string of preceding LoAFs that overlap startTime, I think are more likely to be red-herrings.

Input Delay is only affected by Tasks/LoAFs when the event is already queued on the main thread (and so blocking processingStart).  But startTime is based on hardware timeStamp, and the event can be stuck in hardware, or platform, or browser process... in this case LoAFs aren't actually related.

Once the event is on the main thread, it is scheduled with the HIGHEST priority and there cannot be a case where we somehow do multiple animation frames before deciding to schedule the event processing.

Unless I'm overlooking something ;)

Of course, that is based on today's specifics implementation details of Chromium.  It's already different on Firefox, though they don't report LoAF, anyway.

---

I would personally prefer:
- A single LoAF for "the" frame
- A list of extra LoAF, for Input Delay

...but we try to constrain that list to our best guess of what actually affected the interaction, based on platform and version, and can start with the single most recent LoAF that is `event.startTime <= loaf.endTime < event.processingStart` or something.

But I would also be fine with just a single list, especially in reverse chronological order.

Reply:
> The loaf that intersects with processingStart / processingEnd I would consider "the" LoAF for this Interaction, and it feels uniquely important vs just a list of all of them.

I don't get why that's the important one rather than the longest one?

Like if you had a 200ms input delay (e.g. due to a long rAF) and then 10ms event handler I don't see why the event handler LoAF is "the" LoAF for the interaction as apposed to the one that actually caused the delay?

Though I do agree the event handler one is likely to be the longest one in most cases so maybe a bit of a moot point.

> I would personally prefer:
>
> - A single LoAF for "the" frame
> - A list of extra LoAF, for Input Delay

I definitely prefer a single list. It's all complicated enough and think most people are likely to just pull out the script info rather than do any deep diving in to the other LoAF info anyway. So one list just makes it less complicated.

Reply:
> The one single preceding LoAF might well be useful for debugging input delay.
> 
> However, an long string of preceding LoAFs that overlap startTime, I think are more likely to be red-herrings.

In all of the testing I did for this, I never once saw more than 2 LoAF entries reported in the array. Can there really be  a string of long preceding LoAFs? If so, do you have any suggestions for how I can repro that, because I'd also want to test that scenario.

Reply:
I think that you can only have 1 single "Main thread input delay" LoAF and 1 single "Main thread processing" LoAF, and sometimes those are even the same LoAF.

But you can in theory have variable length extra LoAF for non-main thread input delay.  These LoAFs are red-herring and should just be ignored, but if you naively use startTime as the filter, you will include them.

So the question becomes, how can you force non-main thread input delay?

---

One way is the 300ms non-mobile-viewport tap-delay thing.  You click event will have a really large non-main input delay.  If you happen to have LoAFs on main, those would get reported-- but these aren't the cause of the delay.

Another way on MacOS that is somewhat consistent is to use native drag & drop (highlight and drag text, or drag an image) which causes browser-side jank often delaying event dispatch.  Again, if you happen to also have LoAFs at that time, those would overlap but wouldn't really be the cause.

More generally, you cannot force these types of delays consistently, because that only happens with system jank.  And we try hard not to Jank browser main thread, but it does still happen all the time.

Reply:
I suppose another option could be to keep this as an array, but expose a max of 1 that overlaps with processing and then 1 that overlaps with input delay (the last one, chronologically). So a max length of 2 total.

As a meta comment, at this point I'd rather pick an API shape that is flexible for future additions and unlikely to require breaking changes if we want to add anything. I think there's a higher chance that we later decide that individual LoAF entries doesn't give developers everything they need than if we offer an array of intersecting entries.